### PR TITLE
Fix/default branch issue

### PR DESCRIPTION
**Pull request by felipefoz**
Hi,
I just did the modifications we were talking. It actually took me some time, python is not the my main language, and I've been for a while without programming in it.
Anyways, it turns out that defining the default branch in git is not something deterministic, and after trying some approaches, it ended up in many failing tests. Reading a bit more, I found out that the only way to make sure about the default branch is consulting the remote or even some API.
Instead I decided to use another approach, with a list of possible default branches, it checks if one of these branches are in the list of all branches. There is a corner case where it could have both, and the default turns to be not in first place of the list.

About tests. I couldn't figure out why a test keeps failing. I am pretty sure it is something I did, because it wasn't before, but I couldn't figure out why, could you help on that?
```bash
file: peru\tests\test_plugins.py", line 168, in test_git_plugin_multiple_fetches
    self.assertEqual(output.count("git clone"), 1)
AssertionError: 0 != 1
```
Cheers,

### CODE REVIEW

File: tests/shared.py [hunk 1]
```diff
@@ -186,10 +186,14 @@ def run(self, *command):
 
 
 class GitRepo(Repo):
-    def __init__(self, content_dir):
+    def __init__(self, content_dir, **init_default_branch):
```
Comment (oconnor663) [inline tests/shared.py:189]: The `**` syntax tells Python to collect _any_ keyword arguments into a dictionary called `init_default_branch`. What we'd prefer is to define a single default argument, like this:

```python
def __init__(self, content_dir, init_default_branch=None):
```

Then we can test it like this:

```python
if init_default_branch is not None:
```

There might also be some more specialized syntax we can take advantage of here, see below...

File: tests/shared.py [hunk 2]
```diff
@@ -186,10 +186,14 @@ def run(self, *command):
...
         super().__init__(content_dir)
 
         self.run('git', 'init')
+        # includes option to change the initial default branch
+        if 'default_branch' in init_default_branch:
+            self.run('git', 'symbolic-ref',
+            'HEAD', 'refs/heads/' + init_default_branch['default_branch'])a
```
Comment (oconnor663) [inline tests/shared.py:196]: What's the `a` doing at the end of this line? I feel like that should be a syntax error? That might be your test failure. One way to check this is to introduce some HUGE syntax error and see if you get a different failure. If not, then your current failure is probably a syntax error :)

File: tests/shared.py [hunk 3]
```diff
@@ -186,10 +186,14 @@ def run(self, *command):
 
 
 class GitRepo(Repo):
-    def __init__(self, content_dir):
+    def __init__(self, content_dir, **init_default_branch):
         super().__init__(content_dir)
 
         self.run('git', 'init')
```
Comment (oconnor663) [inline tests/shared.py:192]: I think the best approach will be to use the `--initial-branch` flag here as part of the `git init` command, in the case where `init_default_branch` is set. Python has some special syntax for "include all the elements of this list as positional arguments", so the call might end up looking like:

```python
branch_args = []
if init_default_branch is not None:
    branch_args = ["--branch", init_default_branch]
# Note the * character. This has no effect if the list is empty.
self.run('git', 'init', *branch_args)
```

File: peru/resources/plugins/git/git_plugin.py [hunk 4]
```diff
@@ -167,10 +168,33 @@ def plugin_reup(url, reup):
...
+    By default git used to take master as default branch, but newer versions
+    tend to use main, even trunk or development can be found.
+
+    Args:
+        url (str): url from the target repository to be checked.
+    Returns:
+        str: returns a possible match for the git default branch.
+    """
+    repo_path = clone_if_needed(url)
+    default_branches_list = ['master', 'main', 'trunk', 'development']
```
Comment (oconnor663) [inline peru/resources/plugins/git/git_plugin.py:184]: My personal preference is to just use `master` and `main` here. I think folks with more customized setups should use the explicit `rev` or `reup` fields in their `peru.yaml`.

File: peru/resources/plugins/git/git_plugin.py [hunk 5]
```diff
@@ -167,10 +168,33 @@ def plugin_reup(url, reup):
...
+    for branch in default_branches_list:
+        output = git(
+            'branch', '-a', git_dir=repo_path, capture_output=True).output
+        parse_output = re.search(r'' + re.escape(branch) + '', output)
+        if parse_output:
```
Comment (oconnor663) [inline peru/resources/plugins/git/git_plugin.py:189]: Rather than parsing the text output of `git branch` (which is human-oriented text that might not be perfectly stable), what I'd rather use here is the method suggested at https://stackoverflow.com/questions/5167957/is-there-a-better-way-to-find-out-if-a-local-git-branch-exists. That command has no output, and instead you look at its error code. (`0` means success, i.e. the branch exists. Anything else means failure, i.e. it doesn't exist.) You'll need to use the `checked=False` keyword arg to avoid throwing an exception in that failure case; see the similar examples in the `already_has_ref()` function.

File: peru/resources/plugins/git/git_plugin.py [hunk 6]
```diff
@@ -167,10 +168,33 @@ def plugin_reup(url, reup):
...
+            return branch
+    return 'master'
+
+
 def main():
     URL = os.environ['PERU_MODULE_URL']
-    REV = os.environ['PERU_MODULE_REV'] or 'master'
-    REUP = os.environ['PERU_MODULE_REUP'] or 'master'
+    REV = os.environ['PERU_MODULE_REV'] or git_default_branch(URL)
+    REUP = os.environ['PERU_MODULE_REUP'] or git_default_branch(URL)
```
Comment (oconnor663) [inline peru/resources/plugins/git/git_plugin.py:197]: Since `git_default_branch` is doing nontrivial work, let's call it just once at the top, and capture its return value in a local variable.

**Review by oconnor663 — CHANGES_REQUESTED** [review]
Thanks for doing all this work! And especially for putting up with a less familiar language. I have some requested changes above, and hopefully one of those will help you with the failing test.

**Comment by oconnor663** [comment]
Another strategy we could consider here: I believe when a repo is cloned, the local HEAD is copied from the remote HEAD. So if a repo has both `master` and `main` branches, we could figure out which one to use by just looking at HEAD (`git branch --show-current`) after the clone. I fact, we might not need to special-case `master` and `main` at all with that approach.

However, one issue I see there, is that I think the state of our HEAD could get stale. If we cloned the repo last week, and this week the default branch was changed, would we ever find out without clearing our cache? It looks like `git ls-remote --symref` might be a workaround for that, to let us directly inspect the HEAD of the remote, but I'm not 100% clear on how that works yet.

File: tests/shared.py (hunk 1 above)
Comment (felipefoz) [reply tests/shared.py:189]: Nice, one new thing learned.

File: tests/shared.py (hunk 3 above)
Comment (felipefoz) [reply tests/shared.py:192]: Looks nicer indeed. 
But just a fix, the option would be "--initial-branch" or "-b" for short.

File: tests/shared.py (hunk 2 above)
Comment (felipefoz) [reply tests/shared.py:196]: Rookie mistake, sorry.
Actually I removed this code from there, and it is still failing at the same test, and it is not an error.
I will keep investigating.

File: peru/resources/plugins/git/git_plugin.py (hunk 4 above)
Comment (felipefoz) [reply peru/resources/plugins/git/git_plugin.py:184]: Agreed.

File: peru/resources/plugins/git/git_plugin.py (hunk 5 above)
Comment (felipefoz) [reply peru/resources/plugins/git/git_plugin.py:189]: Ahhhha, that was the code I was looking for. 
Everytime I was using the git wrapper I had this problem. For this reason I haven't gone this route. Let me see what I can come up.

File: peru/resources/plugins/git/git_plugin.py (hunk 6 above)
Comment (felipefoz) [reply peru/resources/plugins/git/git_plugin.py:197]: Good, agreed.

File: peru/resources/plugins/git/git_plugin.py (hunk 5 above)
Comment (felipefoz) [reply peru/resources/plugins/git/git_plugin.py]: It looks better now

**Comment by felipefoz** [comment]
> Another strategy we could consider here: I believe when a repo is cloned, the local HEAD is copied from the remote HEAD. So if a repo has both `master` and `main` branches, we could figure out which one to use by just looking at HEAD (`git branch --show-current`) after the clone. I fact, we might not need to special-case `master` and `main` at all with that approach.
> 
> However, one issue I see there, is that I think the state of our HEAD could get stale. If we cloned the repo last week, and this week the default branch was changed, would we ever find out without clearing our cache? It looks like `git ls-remote --symref` might be a workaround for that, to let us directly inspect the HEAD of the remote, but I'm not 100% clear on how that works yet.
> 
> Btw, it looks like this position of the `--symref` flag is important. So this works (actually it works even if you're not inside a repo, which is nice):
> 
> ```
> git ls-remote --symref https://github.com/buildinspace/peru HEAD
> ```
> 
> But moving the `--symref` flag any farther to the right means we don't get the symref output. That's kind of lame.

I thought of something like this, but that would require changing some strategy in testing, because all the testing considers a local git repository creation (I think), correct me if I am wrong.
For this reason, I thought this approach would solve 99% of the problems.

File: tests/shared.py (hunk 2 above)
Comment (felipefoz) [reply tests/shared.py]: I just removed the function from main() in git_plugin.py and the failures is gone.
Is `def test_git_plugin_multiple_fetches(self)` doing something different that could interfere?
I still do not completely understand the testing functions.

Comment (felipefoz) [reply tests/shared.py]: Somehow, running this function (git_default_branch) before the `plugin_sync` or `plugin_reup` functions is causing the test to fail. 
The only thing that touches another part is `clone_if_needed(),` so it made me think that running this `clone_if_needed() `before is causing something to collapse in the multiple fetches test. Any clues?

File: tests/shared.py [hunk 7]
```diff
@@ -186,10 +186,14 @@ def run(self, *command):
...
-    def __init__(self, content_dir):
+    def __init__(self, content_dir, init_default_branch=None):
         super().__init__(content_dir)
 
-        self.run('git', 'init')
+        branch_args = []
+        if init_default_branch is not None:
+            branch_args = ['--initial-branch', init_default_branch]
+        # Note the * character. This has no effect if the list is empty.
+        self.run('git', 'init', *branch_args)
```
Comment (oconnor663) [inline tests/shared.py:196]: Now that I'm testing this, I'm reminded that our `test.py` script has been spitting out a ton of warnings about how the "default branch is subject to change". Maybe the right thing to do would be to just always pass in `--initial-branch`, to suppress these warnings? So maybe the default value could be `"master"` for now, instead of `None`?

**Comment by oconnor663** [comment]
Ah I see what's happening with `test_git_plugin_multiple_fetches`. It's pretty non-obvious, sorry about that: That test is counting the number of clones and fetches that get executed, by looking for print statements in the plugin output. So because the new code is doing a clone without printing, the count is different from what the test expects. Now that I look at the code, I honestly can't remember why the `clone_and_maybe_print` needs to exist, and I think the easiest thing to do is to just get rid of it and make `clone_if_needed` do the printing instead. Like this:

```diff
diff --git a/peru/resources/plugins/git/git_plugin.py b/peru/resources/plugins/git/git_plugin.py
index 93f0aa4..dab533a 100755
--- a/peru/resources/plugins/git/git_plugin.py
+++ b/peru/resources/plugins/git/git_plugin.py
@@ -40,6 +40,8 @@ def has_clone(url):
 def clone_if_needed(url):
     repo_path = repo_cache_path(url)
     if not has_clone(url):
+        # We look for this print in tests, to count the number of clones we did.
+        print('git clone ' + url)
         git('clone', '--mirror', '--progress', url, repo_path)
     return repo_path
 
@@ -58,12 +60,6 @@ def repo_cache_path(url):
     return os.path.join(CACHE_ROOT, url_hash)
 
 
-def clone_and_maybe_print(url):
-    if not has_clone(url):
-        print('git clone ' + url)
-    return clone_if_needed(url)
-
-
 def git_fetch(url, repo_path):
     print('git fetch ' + url)
     git('fetch', '--prune', git_dir=repo_path)
@@ -88,7 +84,7 @@ def already_has_rev(repo, rev):
 
 
 def checkout_tree(url, rev, dest):
-    repo_path = clone_and_maybe_print(url)
+    repo_path = clone_if_needed(url)
     if not already_has_rev(repo_path, rev):
         git_fetch(url, repo_path)
     # If we just use `git checkout rev -- .` here, we get an error when rev is
```

**Comment by oconnor663** [comment]
> that would require changing some strategy in testing, because all the testing considers a local git repository creation (I think), correct me if I am wrong

When I've played with `ls-remote` by hand, it seems like it works the same way whether or not the remote repo is on the same machine. (So if you're working in an SSH clone, `ls-remote` will make network calls.) If so, does that change the situation?

File: tests/shared.py (hunk 7 above)
Comment (felipefoz) [reply tests/shared.py:196]: Why not?

File: tests/shared.py (hunk 2 above)
Comment (felipefoz) [reply tests/shared.py]: Yay, with the removal of the function `clone_maybe_print()` the error in the test is gone.

**Comment by felipefoz** [comment]
> Ah I see what's happening with `test_git_plugin_multiple_fetches`. It's pretty non-obvious, sorry about that: That test is counting the number of clones and fetches that get executed, by looking for print statements in the plugin output. So because the new code is doing a clone without printing, the count is different from what the test expects. Now that I look at the code, I honestly can't remember why the `clone_and_maybe_print` needs to exist, and I think the easiest thing to do is to just get rid of it and make `clone_if_needed` do the printing instead. Like this:

Now all tests are passing.

**Comment by felipefoz** [comment]
> > that would require changing some strategy in testing, because all the testing considers a local git repository creation (I think), correct me if I am wrong
> 
> When I've played with `ls-remote` by hand, it seems like it works the same way whether or not the remote repo is on the same machine. (So if you're working in an SSH clone, `ls-remote` will make network calls.) If so, does that change the situation?

I can't confirm, because I usually don't work with SSH.
Shall we do some refactoring? or do we leave this for some future bug haunts us?

File: peru/resources/plugins/git/git_plugin.py [hunk 8]
```diff
@@ -167,10 +163,33 @@ def plugin_reup(url, reup):
         print('rev:', output.strip(), file=out_file)
 
 
+def git_default_branch(url) -> str:
+    """
+    With a list of common default branches names, it checks against
+    the available branch list if this possible default branch exists.
+    By default git used to take master as default branch, but newer versions
+    tend to use main, even trunk or development can be found.
```
Comment (oconnor663) [inline peru/resources/plugins/git/git_plugin.py:171]: We could update this comment, since we've shortened the list. But I can also do that myself after landing, no worries.

Comment (felipefoz) [reply peru/resources/plugins/git/git_plugin.py:171]: I do it, I am also going to simplify the function a little bit.

**Comment by oconnor663** [comment]
This looks good to me. CI is green, so as soon as you're satisfied with the cleanup I'll go ahead and land it. As far as the specific strategy we use for repos that have both `"main"` and `"master"`, I'm fine with sticking with the current simple approach. If we do something more complicated with `ls-remote`, that increases the chance that some unexpected change on the server side that we can't control will "break" us. The whole default branch name situation in Git is inherently unstable anyway. I'd prefer that users who really care about this case just make their preferences explicit in `peru.yaml`.

**Comment by felipefoz** [comment]
> This looks good to me. CI is green, so as soon as you're satisfied with the cleanup I'll go ahead and land it. As far as the specific strategy we use for repos that have both `"main"` and `"master"`, I'm fine with sticking with the current simple approach. If we do something more complicated with `ls-remote`, that increases the chance that some unexpected change on the server side that we can't control will "break" us. The whole default branch name situation in Git is inherently unstable anyway. I'd prefer that users who really care about this case just make their preferences explicit in `peru.yaml`.

Agreed. I included this comment in code. If user is using a different approach, then should use the rev option.

**Comment by felipefoz** [comment]
Great. I think I am done with this one. CI kept green!

**Comment by oconnor663** [comment]
Released as version 1.3.0.
